[NextJS]Style guide V1 (#232)
Wolox/tech-guides: frontend/NextJs/style-guide.md

# Next.js Style Guide

## Table of Contents

1. [Basic Rules](#basic-rules)
1. [Folder Structure](#folder-structure)
1. [Fetching Data](#fetching-data)
1. [Api Routes](#naming)

> 1. [SEO](#mixins)

Reviewer: These links should be updated

## Basic Rules

- If you are not using our Next.js Boostrap, make sure your the next.config.js
  file include pageExtensions as follows
  ```js
  module.exports = {
    pageExtensions: ['page.tsx', 'api.ts'],
  }
  ```
- Only include one index.page.tsx component per page.
- Always use JSX syntax.

## Folder Structure

```
└───locales //folder to i18n per page
│ └─── es
│     └─── home.json
│     └─── myscreenpage.json
│     └───etc
│ └─── en
│     └─── home.json
│     └─── myscreenpage.json
│     └───etc
│
└───components
│ └───baseComponents
│     └───Input
│     └───Text
│     └───Button
│     └───etc
│
└───pages

> │   | index.page.tsx // root or hoem page.

Reviewer: ```suggestion
│   | index.page.tsx // root or home page.
```

│   └───api
│   │   └───MyServerlessApi
│   │       | index.api.tsx
│   └───MyScreenPage
│       | components/
│       | constants.ts
│       | index.page.tsx
│       | styles.module.scss //only if you are using css modules.
│       | utils.ts
│       | assets // Screen specific app assets
│
└─── public
│ └─── assets
|
└───config
    | api.ts
└───constants
└─── store (Redux/Contex/Recoil)
│   | store.ts
│   └───myReducer
│       | actions.ts
│       | reducer.ts
│       | selectors.ts
└─── hooks
└─── types
└─── services
    | MyService.ts
└───utils
│   index.ts
```

## Fetching Data

Next.js allows you to fetch data in different ways, depending on your needs and
preferences.

> - **getStaticProps**: The getStaticProps function is executed on the server side

Reviewer: How about this? The function description may result a little bit redundant if we name the function after it's "section title"

```suggestion
- **getStaticProps**: This function is executed on the server side
```

Reviewer: The same applies to other functions

  is compiled, and allows you to retrieve data from an external API or any other

>   source and pass it as properties to the component. This function is util when

Reviewer: I think that "util" as an adjective does not exist in english (idk if "util" may be used as a noun such as the "utils" directory on some projects though)
```suggestion
  source and pass it as properties to the component. This function is useful when
```

>   we need to have some pages prebuilt like **blogs**, or **lading

Reviewer: ```suggestion
  we need to have some pages prebuilt like **blogs**, or **landing
```

  pages**.<br /><br />

  ```js
  function MyComponent(props) {
    return (
      // Use the retrieved data as component props
      <div>{props.data}</div>
    )
  }

  export async function getStaticProps() {
    // Retrieve data from the external API using the Fetch API
    const res = await fetch('https://my-api.com/datos')
    const data = await res.json()

    // Pass the data as props to the component
    return {props: {data}}
  }
  ```

- **getServerSideProps**: The getServerSideProps function, which is executed on
  the server side every time a page is accessed, to fetch data and pass it to

>   the component as properties. yop can also get the query parameters in the URL

Reviewer: ```suggestion
  the component as properties. you can also get the query parameters in the URL
```

  as request parameters.<br /><br />

  ```js
  function MyComponent(props) {
    return (
      // Use the retrieved data as component props
      <div>{props.data}</div>
    )
  }

  export async function getServerSideProps(context) {

>     // Retrieve data from theexternal API

Reviewer: ```suggestion
    // Retrieve data from the external API
```

data. This endpoint could then be used to perform database queries or call
external APIs to retrieve the necessary data and return it to the client in a
JSON. Here is an example of an API route in Next.js that returns a list of users
in JSON format: // pages/api/users.js

```js
import users from './data/users.json'

export default (req, res) => {
  res.status(200).json(users)
}
```

This API route exports a default function that is invoked when a request is made
to the /api/users endpoint. The function uses the res object to send a JSON
response to the client with a list of users. You can then make a request to this
endpoint from your front-end code using a tool like fetch. For example:

```js
fetch('/api/users')
  .then(res => res.json())
  .then(data => console.log(data))
```

This will send a GET request to the /api/users endpoint and log the response
data to the console.

You can read more examples
[here](https://github.com/vercel/next.js/tree/canary/examples/api-routes-rest)
You can read the docs [here](https://nextjs.org/docs/api-routes/introduction)

## SEO

In Next.js you can improve the SEO of the websites using some

> [pre-rendering modes](#fetching-data) this modes will we could to get a document

Reviewer: I think that the redaction here is a little bit messy, don't know if I'm getting the right idea, how about?
```suggestion
[pre-rendering modes](#fetching-data) these modes will generate the application
```

Reviewer 2: Hi @MMarussi! what about `these modes will let you get an optimized HTML document`?

Reviewer: Perfect!

HTML optimized for the web crawler and indexers.

- **Prerendering modes**:

>   - Static Generation: HTML is only generating during build time.
> 
>   - Server-side Rendering: HTML is generating on each reques.<br /><br />

Reviewer: ```suggestion
  - Static Generation: HTML is only generated during build time.

  - Server-side Rendering: HTML is generated on each request.<br /><br />
```

- **Head Component**: Next.js provides a built-in component called ‘Head’, which
  helps to append different elements such as title tag, meta tag etc. to the
  <head> part of the document. You can import it from next/head.<br /><br />

  ```js
  import Head from 'next/head'

  function IndexPage() {
    return (
      <div>
        <Head>
          <title>My page title</title>
          <meta name="description" content="This is my page description" />{' '}
        </Head>
        <p>Hello world!</p>{' '}
      </div>
    )
  }

  export default IndexPage
  ```

>   Head Component makes sure all of your pages include important details like
>   title, description etc. that need to get rendered into your page. Next.js
>   recognises this metadata and lift it to the right location in your HTML
>   document when the page is being rendered. It helps in improving SEO.

Reviewer: How about this
```suggestion
  Head Component helps us to improve SEO by making sure all of your pages include important details like
  title, description, etc. that need to get rendered into your page. Next.js
  recognises this metadata and lift it to the right location in your HTML
  document when the page is being rendered.
```
